src/server/frontend/render.js

```diff
@@ -33,48 +32,44 @@ const fetchComponentDataAsync = async (dispatch, renderProps) => {
   await Promise.all(promises);
 };
 
-const getAppHtml = (store, renderProps) =>
-  ReactDOMServer.renderToString(
+const getAppContainer = (state, store, renderProps) =>
+  <div id="app" data-initial-state={JSON.stringify(state)}>
     <Provider store={store}>
       <IntlProvider>
         <RouterContext {...renderProps} />
       </IntlProvider>
     </Provider>
-  );
+  </div>;
 
-const getScriptHtml = (state, headers, hostname, appJsFilename) =>
-  // Note how app state is serialized. JSON.stringify is anti-pattern.
-  // https://github.com/yahoo/serialize-javascript#user-content-automatic-escaping-of-html-characters
+const getScripts = (appJsFilename) =>
   // Note how we use cdn.polyfill.io, en is default, but can be changed later.
-  `
+  <div>
     <script src="https://cdn.polyfill.io/v2/polyfill.min.js?features=Intl.~locale.en"></script>
-    <script>
-      window.__INITIAL_STATE__ = ${serialize(state)};
-    </script>
-    <script src="${appJsFilename}"></script>
-  `;
+    <script src={appJsFilename}></script>
+  </div>;
 
-const renderPage = (store, renderProps, req) => {
+const renderPage = (store, renderProps) => {
   const state = store.getState();
-  const { headers, hostname } = req;
-  const appHtml = getAppHtml(store, renderProps);
+  const appContainer = getAppContainer(state, store, renderProps);
   const helmet = Helmet.rewind();
   const {
     styles: { app: appCssFilename },
     javascript: { app: appJsFilename }
   } = webpackIsomorphicTools.assets();
-  const scriptHtml = getScriptHtml(state, headers, hostname, appJsFilename);
+  const scripts = getScripts(appJsFilename);
   if (!config.isProduction) {
     webpackIsomorphicTools.refresh();
   }
   const docHtml = ReactDOMServer.renderToStaticMarkup(
```

Comment:
this will screw server-side renderig up, as React will lose it's ids and checksum (`renderToString` keeps them, `renderToStaticMarkup` doesn't)